contrib/dynamic_embedding/src/tde/ps.cpp

```diff
@@ -25,9 +25,12 @@ c10::intrusive_ptr<FetchHandle> PS::Fetch(
   if (cache_ids_to_fetch_or_evict_.empty()) {
     return c10::make_intrusive<FetchHandle>(time, c10::intrusive_ptr<PS>());
   }
-  fetch_notifications_.emplace_back(time, c10::make_intrusive<Notification>());
-  c10::intrusive_ptr<Notification> notification =
-      fetch_notifications_.back().second;
+  c10::intrusive_ptr<Notification> notification;
+  {
+    std::unique_lock<std::mutex> lock_fetch(fetch_notifications_mutex_);
```

Comment:
Why is this lock needed? Line 20 grabs the lock mu_, so this function should be thread safe?

Reply:
Although Fetch grabs the mu_ lock, the notification is accessed in SyncFetch without the lock (mu_), so the access to notification is not thread-safe. Here, a finer-grained lock is chosen to independently protect the notification instead of adding the mu_ lock during the SyncFetch process. The reason for this will be explained below.

Reply:
Do we even need mu_ lock then? Is that something we can get rid of? Could be a follow up pr

Reply:
Based on my understanding, the primary purpose of mu_ lock is to ensure that the calls to Fetch and Evict are mutually exclusive, as variables such as cache_ids_to_fetch_or_evict_ and shards_ are shared between them. Although the current implementation in the Python side actually guarantees that they won't be called simultaneously, the way they are called from the Python side may change in the future. Therefore, the mu_ lock might still be necessary.

Reply:
Additionally, we've encountered performance issues .  The main problem is that both fetch and evict access data by reading or writing single tensors, which results in a large number of small GPU operators (on the same order of magnitude as the number of keys).  for example https://github.com/pytorch/torchrec/blob/main/contrib/dynamic_embedding/src/tde/ps.cpp#L58
By merging these into a larger batch operator, we can achieve a significant performance improvement. We'll clean up the code and submit it as a new mr later.